File: test/task_queue_test.ts
import { assertEquals } from "@std/assert";
import * as tq from "../src/task-queue.ts";

async function makeTempDir(): Promise<string> {
  return await Deno.realPath(
    await Deno.makeTempDir({ prefix: "jackops-tq-" }),
  );
}

const TASK_A = {
  id: "task-001",
  summary: "Add auth module",
  description: "Implement JWT authentication",
  files: ["src/auth.ts"],
  acceptance: ["tokens are validated", "expired tokens rejected"],
  createdBy: "planner",
};

const TASK_B = {
  id: "task-002",
  summary: "Add tests",
  description: "Write tests for auth module",
  depends_on: ["task-001"],
  createdBy: "planner",
};

Deno.test("init creates task directories", async () => {
  const dir = await makeTempDir();
  try {
    await tq.init(dir);
    for (const state of tq.TASK_STATES) {
      const stat = await Deno.stat(`${dir}/.jackops/tasks/${state}`);
      assertEquals(stat.isDirectory, true);
    }
  } finally {
    await Deno.remove(dir, { recursive: true });
  }
});

Deno.test("init is idempotent", async () => {
  const dir = await makeTempDir();
  try {
    await tq.init(dir);
    await tq.init(dir);
    for (const state of tq.TASK_STATES) {
      const stat = await Deno.stat(`${dir}/.jackops/tasks/${state}`);
      assertEquals(stat.isDirectory, true);
    }
  } finally {
    await Deno.remove(dir, { recursive: true });
  }
});

Deno.test("create writes task to pending", async () => {
  const dir = await makeTempDir();
  try {
    await tq.init(dir);
    const task = await tq.create(dir, TASK_A);
    assertEquals(task.id, "task-001");
    assertEquals(task.retries, 0);

    const entry = await tq.get(dir, "task-001");
    assertEquals(entry !== null, true);
    assertEquals(entry!.state, "pending");
    assertEquals(entry!.task.summary, "Add auth module");
    assertEquals(entry!.task.files, ["src/auth.ts"]);
  } finally {
    await Deno.remove(dir, { recursive: true });
  }
});

Deno.test("claim moves task to current with assignee", async () => {
  const dir = await makeTempDir();
  try {
    await tq.init(dir);
    await tq.create(dir, TASK_A);

    const task = await tq.claim(dir, "task-001", "impl-1");
    assertEquals(task.assignee, "impl-1");

    const entry = await tq.get(dir, "task-001");
    assertEquals(entry!.state, "current");
    assertEquals(entry!.task.assignee, "impl-1");
  } finally {
    await Deno.remove(dir, { recursive: true });
  }
});

Deno.test("complete moves task to complete", async () => {
  const dir = await makeTempDir();
  try {
    await tq.init(dir);
    await tq.create(dir, TASK_A);
    await tq.claim(dir, "task-001", "impl-1");

    await tq.complete(dir, "task-001");

    const entry = await tq.get(dir, "task-001");
    assertEquals(entry!.state, "complete");
    assertEquals(entry!.task.assignee, "impl-1");
  } finally {
    await Deno.remove(dir, { recursive: true });
  }
});

Deno.test("review moves current task to review", async () => {
  const dir = await makeTempDir();
  try {
    await tq.init(dir);
    await tq.create(dir, TASK_A);
    await tq.claim(dir, "task-001", "impl-1");

    await tq.review(dir, "task-001");

    const entry = await tq.get(dir, "task-001");
    assertEquals(entry!.state, "review");
    assertEquals(entry!.task.assignee, "impl-1");
  } finally {
    await Deno.remove(dir, { recursive: true });
  }
});

Deno.test("approve moves reviewed task to complete", async () => {
  const dir = await makeTempDir();
  try {
    await tq.init(dir);
    await tq.create(dir, TASK_A);
    await tq.claim(dir, "task-001", "impl-1");
    await tq.review(dir, "task-001");

    await tq.approve(dir, "task-001");

    const entry = await tq.get(dir, "task-001");
    assertEquals(entry!.state, "complete");
    assertEquals(entry!.task.assignee, "impl-1");
  } finally {
    await Deno.remove(dir, { recursive: true });
  }
});

Deno.test("reject moves reviewed task to rejected with feedback", async () => {
  const dir = await makeTempDir();
  try {
    await tq.init(dir);
    await tq.create(dir, TASK_A);
    await tq.claim(dir, "task-001", "impl-1");
    await tq.review(dir, "task-001");

    const task = await tq.reject(dir, "task-001", "missing error handling");

    assertEquals(task.feedback, "missing error handling");

    const entry = await tq.get(dir, "task-001");
    assertEquals(entry!.state, "rejected");
    assertEquals(entry!.task.feedback, "missing error handling");
  } finally {
    await Deno.remove(dir, { recursive: true });
  }
});

Deno.test("retry moves rejected task back to pending", async () => {
  const dir = await makeTempDir();
  try {
    await tq.init(dir);
    await tq.create(dir, TASK_A);
    await tq.claim(dir, "task-001", "impl-1");
    await tq.review(dir, "task-001");
    await tq.reject(dir, "task-001", "needs work");

    const task = await tq.retry(dir, "task-001");
    assertEquals(task.retries, 1);
    assertEquals(task.assignee, undefined);
    assertEquals(task.feedback, undefined);

    const entry = await tq.get(dir, "task-001");
    assertEquals(entry!.state, "pending");
  } finally {
    await Deno.remove(dir, { recursive: true });
  }
});

Deno.test("list returns all tasks across states", async () => {
  const dir = await makeTempDir();
  try {
    await tq.init(dir);
    await tq.create(dir, TASK_A);
    await tq.create(dir, TASK_B);
    await tq.claim(dir, "task-001", "impl-1");

    const all = await tq.list(dir);
    assertEquals(all.length, 2);

    const pending = await tq.list(dir, "pending");
    assertEquals(pending.length, 1);
    assertEquals(pending[0].task.id, "task-002");

    const current = await tq.list(dir, "current");
    assertEquals(current.length, 1);
    assertEquals(current[0].task.id, "task-001");
  } finally {
    await Deno.remove(dir, { recursive: true });
  }
});

Deno.test("list returns empty for uninitialized dir", async () => {
  const dir = await makeTempDir();
  try {
    const all = await tq.list(dir);
    assertEquals(all.length, 0);
  } finally {
    await Deno.remove(dir, { recursive: true });
  }
});

Deno.test("get returns null for nonexistent task", async () => {
  const dir = await makeTempDir();
  try {
    await tq.init(dir);
    const entry = await tq.get(dir, "nonexistent");
    assertEquals(entry, null);
  } finally {
    await Deno.remove(dir, { recursive: true });
  }
});

Deno.test("counts returns task counts by state", async () => {
  const dir = await makeTempDir();
  try {
    await tq.init(dir);
    await tq.create(dir, TASK_A);
    await tq.create(dir, TASK_B);
    await tq.claim(dir, "task-001", "impl-1");

    const c = await tq.counts(dir);
    assertEquals(c.pending, 1);
    assertEquals(c.current, 1);
    assertEquals(c.review, 0);
    assertEquals(c.complete, 0);
    assertEquals(c.rejected, 0);
  } finally {
    await Deno.remove(dir, { recursive: true });
  }
});

Deno.test("ready filters by dependency completion", async () => {
  const dir = await makeTempDir();
  try {
    await tq.init(dir);
    await tq.create(dir, TASK_A);
    await tq.create(dir, TASK_B); // depends on task-001

    // task-001 is pending, so task-002 is not ready
    let readyTasks = await tq.ready(dir);
    assertEquals(readyTasks.length, 1);
    assertEquals(readyTasks[0].id, "task-001");

    // Complete task-001
    await tq.claim(dir, "task-001", "impl-1");
    await tq.complete(dir, "task-001");

    // Now task-002 is ready
    readyTasks = await tq.ready(dir);
    assertEquals(readyTasks.length, 1);
    assertEquals(readyTasks[0].id, "task-002");
  } finally {
    await Deno.remove(dir, { recursive: true });
  }
});

Deno.test("ready returns tasks with no dependencies", async () => {
  const dir = await makeTempDir();
  try {
    await tq.init(dir);
    await tq.create(dir, TASK_A); // no depends_on

    const readyTasks = await tq.ready(dir);
    assertEquals(readyTasks.length, 1);
    assertEquals(readyTasks[0].id, "task-001");
  } finally {
    await Deno.remove(dir, { recursive: true });
  }
});

Deno.test("full lifecycle: create -> claim -> review -> reject -> retry -> claim -> review -> approve", async () => {
  const dir = await makeTempDir();
  try {
    await tq.init(dir);
    await tq.create(dir, TASK_A);

    // First attempt
    await tq.claim(dir, "task-001", "impl-1");
    await tq.review(dir, "task-001");
    await tq.reject(dir, "task-001", "missing tests");

    // Retry
    const retried = await tq.retry(dir, "task-001");
    assertEquals(retried.retries, 1);

    // Second attempt
    await tq.claim(dir, "task-001", "impl-2");
    await tq.review(dir, "task-001");
    await tq.approve(dir, "task-001");

    const entry = await tq.get(dir, "task-001");
    assertEquals(entry!.state, "complete");
    assertEquals(entry!.task.assignee, "impl-2");
    assertEquals(entry!.task.retries, 1);
  } finally {
    await Deno.remove(dir, { recursive: true });
  }
});

File: src/taskqueue/taskqueue_test.go
package taskqueue

import (
	"os"
	"path/filepath"
	"testing"

	"github.com/AdriaanVE/jack-in/src/domain"
)

func tmpBase(t *testing.T) string {
	t.Helper()
	return t.TempDir()
}

func TestInitCreatesDirectories(t *testing.T) {
	base := tmpBase(t)
	if err := Init(base); err != nil {
		t.Fatal(err)
	}
	for _, s := range domain.TaskStates {
		dir := filepath.Join(base, tasksDir, string(s))
		info, err := os.Stat(dir)
		if err != nil {
			t.Fatalf("expected dir %s to exist: %v", s, err)
		}
		if !info.IsDir() {
			t.Fatalf("expected %s to be a directory", s)
		}
	}
}

func TestAddAndGet(t *testing.T) {
	base := tmpBase(t)
	Init(base)

	task, err := Add(base, "Fix login bug", "Users can't log in with SSO")
	if err != nil {
		t.Fatal(err)
	}
	if task.Summary != "Fix login bug" {
		t.Fatalf("expected summary 'Fix login bug', got %q", task.Summary)
	}
	if task.Description != "Users can't log in with SSO" {
		t.Fatalf("expected description, got %q", task.Description)
	}
	if task.Retries != 0 {
		t.Fatalf("expected retries=0, got %d", task.Retries)
	}

	entry, err := Get(base, task.ID)
	if err != nil {
		t.Fatal(err)
	}
	if entry == nil {
		t.Fatal("expected task entry, got nil")
	}
	if entry.State != domain.StatePending {
		t.Fatalf("expected state pending, got %s", entry.State)
	}
	if entry.Task.Summary != "Fix login bug" {
		t.Fatalf("expected summary match, got %q", entry.Task.Summary)
	}
}

func TestAddDefaultsDescriptionToSummary(t *testing.T) {
	base := tmpBase(t)
	Init(base)

	task, err := Add(base, "Quick fix", "")
	if err != nil {
		t.Fatal(err)
	}
	if task.Description != "Quick fix" {
		t.Fatalf("expected description to default to summary, got %q", task.Description)
	}
}

func TestGetNotFound(t *testing.T) {
	base := tmpBase(t)
	Init(base)

	entry, err := Get(base, "nonexistent")
	if err != nil {
		t.Fatal(err)
	}
	if entry != nil {
		t.Fatalf("expected nil for nonexistent task, got %+v", entry)
	}
}

func TestListByState(t *testing.T) {
	base := tmpBase(t)
	Init(base)
	Add(base, "Task A", "")
	Add(base, "Task B", "")

	entries, err := List(base, domain.StatePending)
	if err != nil {
		t.Fatal(err)
	}
	if len(entries) != 2 {
		t.Fatalf("expected 2 pending tasks, got %d", len(entries))
	}

	entries, err = List(base, domain.StateCurrent)
	if err != nil {
		t.Fatal(err)
	}
	if len(entries) != 0 {
		t.Fatalf("expected 0 current tasks, got %d", len(entries))
	}
}

func TestListAll(t *testing.T) {
	base := tmpBase(t)
	Init(base)
	Add(base, "Task A", "")
	Add(base, "Task B", "")

	entries, err := ListAll(base)
	if err != nil {
		t.Fatal(err)
	}
	if len(entries) != 2 {
		t.Fatalf("expected 2 total tasks, got %d", len(entries))
	}
}

func TestCounts(t *testing.T) {
	base := tmpBase(t)
	Init(base)
	Add(base, "Task A", "")
	Add(base, "Task B", "")
	Add(base, "Task C", "")

	c, err := Counts(base)
	if err != nil {
		t.Fatal(err)
	}
	if c.Pending != 3 {
		t.Fatalf("expected 3 pending, got %d", c.Pending)
	}
	if c.Current != 0 || c.Review != 0 || c.Complete != 0 || c.Rejected != 0 {
		t.Fatalf("expected all other counts=0, got %+v", c)
	}
}

func TestFullLifecycle(t *testing.T) {
	base := tmpBase(t)
	Init(base)

	task, _ := Add(base, "Implement feature", "")

	// Claim (pending -> current)
	claimed, err := Claim(base, task.ID, "scout")
	if err != nil {
		t.Fatal(err)
	}
	if claimed.Assignee != "scout" {
		t.Fatalf("expected assignee 'scout', got %q", claimed.Assignee)
	}
	entry, _ := Get(base, task.ID)
	if entry.State != domain.StateCurrent {
		t.Fatalf("expected current, got %s", entry.State)
	}

	// Review (current -> review)
	_, err = Review(base, task.ID)
	if err != nil {
		t.Fatal(err)
	}
	entry, _ = Get(base, task.ID)
	if entry.State != domain.StateReview {
		t.Fatalf("expected review, got %s", entry.State)
	}

	// Approve (review -> complete)
	_, err = Approve(base, task.ID)
	if err != nil {
		t.Fatal(err)
	}
	entry, _ = Get(base, task.ID)
	if entry.State != domain.StateComplete {
		t.Fatalf("expected complete, got %s", entry.State)
	}
}

func TestRejectAndRetry(t *testing.T) {
	base := tmpBase(t)
	Init(base)

	task, _ := Add(base, "Write docs", "")
	Claim(base, task.ID, "coder")
	Review(base, task.ID)

	// Reject (review -> rejected)
	rejected, err := Reject(base, task.ID, "missing examples")
	if err != nil {
		t.Fatal(err)
	}
	if rejected.Feedback != "missing examples" {
		t.Fatalf("expected feedback, got %q", rejected.Feedback)
	}
	entry, _ := Get(base, task.ID)
	if entry.State != domain.StateRejected {
		t.Fatalf("expected rejected, got %s", entry.State)
	}

	// Retry (rejected -> current, keeping assignee and feedback for iteration)
	retried, err := Retry(base, task.ID)
	if err != nil {
		t.Fatal(err)
	}
	if retried.Retries != 1 {
		t.Fatalf("expected retries=1, got %d", retried.Retries)
	}
	if retried.Assignee != "coder" {
		t.Fatalf("expected assignee kept as 'coder', got %q", retried.Assignee)
	}
	if retried.Feedback != "missing examples" {
		t.Fatalf("expected feedback kept, got %q", retried.Feedback)
	}
	entry, _ = Get(base, task.ID)
	if entry.State != domain.StateCurrent {
		t.Fatalf("expected current after retry, got %s", entry.State)
	}
}

func TestRetryMaxLimit(t *testing.T) {
	base := tmpBase(t)
	Init(base)

	task, _ := Add(base, "Flaky task", "")
	Claim(base, task.ID, "coder")

	// Retry MaxRetries times (should all succeed)
	for i := 0; i < MaxRetries; i++ {
		Review(base, task.ID)
		Reject(base, task.ID, "try again")
		_, err := Retry(base, task.ID)
		if err != nil {
			t.Fatalf("retry %d should succeed, got error: %v", i+1, err)
		}
	}

	// One more retry should fail
	Review(base, task.ID)
	Reject(base, task.ID, "final rejection")
	_, err := Retry(base, task.ID)
	if err != ErrMaxRetriesExceeded {
		t.Fatalf("expected ErrMaxRetriesExceeded, got: %v", err)
	}

	// Task should stay in rejected
	entry, _ := Get(base, task.ID)
	if entry.State != domain.StateRejected {
		t.Fatalf("expected rejected after max retries, got %s", entry.State)
	}
}

func TestUnclaim(t *testing.T) {
	base := tmpBase(t)
	Init(base)

	task, _ := Add(base, "Task X", "")
	Claim(base, task.ID, "worker1")

	unclaimed, err := Unclaim(base, task.ID)
	if err != nil {
		t.Fatal(err)
	}
	if unclaimed.Assignee != "" {
		t.Fatalf("expected assignee cleared, got %q", unclaimed.Assignee)
	}
	entry, _ := Get(base, task.ID)
	if entry.State != domain.StatePending {
		t.Fatalf("expected pending after unclaim, got %s", entry.State)
	}
}

func TestReadyRespectsDependencies(t *testing.T) {
	base := tmpBase(t)
	Init(base)

	taskA, _ := Add(base, "Task A", "")
	taskB, _ := Add(base, "Task B", "")

	// Manually add depends_on to taskB
	entry, _ := Get(base, taskB.ID)
	entry.Task.DependsOn = []string{taskA.ID}
	writeTask(base, domain.StatePending, &entry.Task)

	// Only A should be ready (B depends on A)
	ready, err := Ready(base)
	if err != nil {
		t.Fatal(err)
	}
	if len(ready) != 1 {
		t.Fatalf("expected 1 ready task, got %d", len(ready))
	}
	if ready[0].ID != taskA.ID {
		t.Fatalf("expected task A to be ready, got %s", ready[0].ID)
	}

	// Complete A, now B should be ready
	Claim(base, taskA.ID, "w1")
	Review(base, taskA.ID)
	Approve(base, taskA.ID)

	ready, err = Ready(base)
	if err != nil {
		t.Fatal(err)
	}
	if len(ready) != 1 {
		t.Fatalf("expected 1 ready task after completing A, got %d", len(ready))
	}
	if ready[0].ID != taskB.ID {
		t.Fatalf("expected task B to be ready, got %s", ready[0].ID)
	}
}

func TestReadyMatchesByIDAndSummary(t *testing.T) {
	base := tmpBase(t)
	Init(base)

	taskA, _ := Add(base, "Build API", "")
	// Depend by summary instead of ID
	taskB, _ := Add(base, "Test API", "")
	entry, _ := Get(base, taskB.ID)
	entry.Task.DependsOn = []string{"Build API"}
	writeTask(base, domain.StatePending, &entry.Task)

	// B blocked
	ready, _ := Ready(base)
	if len(ready) != 1 || ready[0].ID != taskA.ID {
		t.Fatalf("expected only A ready, got %v", ready)
	}

	// Complete A
	Claim(base, taskA.ID, "w1")
	Review(base, taskA.ID)
	Approve(base, taskA.ID)

	// B unblocked via summary match
	ready, _ = Ready(base)
	if len(ready) != 1 || ready[0].ID != taskB.ID {
		t.Fatalf("expected B ready after summary match, got %v", ready)
	}
}

func TestSeedSkipsDuplicates(t *testing.T) {
	base := tmpBase(t)

	seeds := []TaskSeed{
		{Summary: "Task One", Description: "First"},
		{Summary: "Task Two", Description: "Second"},
	}

	n, err := Seed(base, seeds)
	if err != nil {
		t.Fatal(err)
	}
	if n != 2 {
		t.Fatalf("expected 2 seeded, got %d", n)
	}

	// Seed again — should skip both
	n, err = Seed(base, seeds)
	if err != nil {
		t.Fatal(err)
	}
	if n != 0 {
		t.Fatalf("expected 0 seeded on re-seed, got %d", n)
	}

	entries, _ := ListAll(base)
	if len(entries) != 2 {
		t.Fatalf("expected 2 total tasks, got %d", len(entries))
	}
}

func TestSeedResolvesDependencies(t *testing.T) {
	base := tmpBase(t)

	seeds := []TaskSeed{
		{Summary: "Foundation"},
		{Summary: "Feature", DependsOn: []string{"Foundation"}},
	}

	_, err := Seed(base, seeds)
	if err != nil {
		t.Fatal(err)
	}

	entries, _ := List(base, domain.StatePending)
	var feature *Task
	var foundation *Task
	for i := range entries {
		if entries[i].Task.Summary == "Feature" {
			feature = &entries[i].Task
		}
		if entries[i].Task.Summary == "Foundation" {
			foundation = &entries[i].Task
		}
	}
	if feature == nil || foundation == nil {
		t.Fatal("expected both tasks to exist")
	}
	if len(feature.DependsOn) != 1 || feature.DependsOn[0] != foundation.ID {
		t.Fatalf("expected depends_on resolved to foundation ID %s, got %v", foundation.ID, feature.DependsOn)
	}
}
